test/streamparse/test_bolt.py

```diff
@@ -0,0 +1,298 @@
+    @patch.object(bolt.Bolt, '_run')
+    def test_auto_fail(self, _run_mock, fail_mock):
+        tup = ipc.Tuple(14, 'some_spout', 'default', 'some_bolt', [1, 2, 3])
+        self.bolt._current_tups = [tup]
+        # Make sure _run raises an exception
+        def raiser(): # lambdas can't raise
+            raise Exception('borkt')
+        _run_mock.side_effect = raiser
+
+        # test auto-fail on (the default)
+        self.bolt.run()
+        fail_mock.assert_called_with(tup)
+        fail_mock.reset_mock()
+
+        # test auto-fail off
+        self.bolt.auto_fail = False
+        self.bolt.run()
+        self.assertFalse(fail_mock.called)
+
+
+@patch('streamparse.bolt.send_message', new=lambda *a: None)
+@patch('streamparse.base.send_message', new=lambda *a: None)
+class BatchingBoltTests(unittest.TestCase):
+
+    def setUp(self):
+        # mock seconds between batches to speed the tests up
+        self._orig_secs = bolt.BatchingBolt.secs_between_batches
+
+        bolt.BatchingBolt.secs_between_batches = 0.05
+        self.bolt = bolt.BatchingBolt()
+        self.bolt.initialize({}, {})
+
+        # Mock read_tuple and manually since it all needs to be mocked
+        self.tups = [
+            ipc.Tuple(14, 'some_spout', 'default', 'some_bolt', [1, 2, 3]),
+            ipc.Tuple(15, 'some_spout', 'default', 'some_bolt', [4, 5, 6]),
+            ipc.Tuple(16, 'some_spout', 'default', 'some_bolt', [7, 8, 9]),
+        ]
+        self._orig_read_tuple = bolt.read_tuple
+        tups_cycle = itertools.cycle(self.tups)
+        bolt.read_tuple = lambda: next(tups_cycle)
+
+    def tearDown(self):
+        # undo the mocking
+        bolt.BatchingBolt.secs_between_batches = self._orig_secs
+        bolt.read_tuple = self._orig_read_tuple
+
+    @patch.object(bolt.BatchingBolt, 'process_batch')
+    def test_batching(self, process_batch_mock):
+        # Add a bunch of tuples
+        for __ in range(3):
+            self.bolt._run()
+
+        # Wait a bit, and see if process_batch was called
+        time.sleep(0.1)
+        process_batch_mock.assert_called_with(None, self.tups[:3])
+
+    @patch.object(bolt.BatchingBolt, 'process_batch')
+    def test_group_key(self, process_batch_mock):
+        # Change the group key to even/odd grouping
+        self.bolt.group_key = lambda t: sum(t.values) % 2
+
+        # Add a bunch of tuples
+        for __ in range(3):
+            self.bolt._run()
+
+        # Wait a bit, and see if process_batch was called correctly
+        time.sleep(0.1)
+        process_batch_mock.assert_has_calls([
+            mock.call(0, [self.tups[0], self.tups[2]]),
+            mock.call(1, [self.tups[1]]),
+        ], any_order=True)
+
+    def _test_exception_handling(self):
+        # Make sure the exception gets from the worker thread to the main
+        for i in range(1):
+            self.bolt._run()
+        self.assertRaises(NotImplementedError, lambda: time.sleep(0.1))
```

Comment:
@dan-blanchard note that this method doesn't run when calling `unittest.main()` on the file. If you rename it to `test_exception_handling()` , it will run and fail